scripts/file-entities.php

```diff
@@ -365,3 +309,23 @@ function writeEntity( $file , Entity $ent )
 
     fwrite( $file , $line );
 }
+
+function realpain( string $path , bool $touch = false ) : string
+{
+    // pain is real
+
+    // care for external XML tools (realpath everywhere)
+    // care for Windows builds (foward slashes everywhere)
+    // avoid `cd` and chdir() like the plague
+
+    $path = str_replace( "\\" , '/' , $path );
+    if ( $touch && ! file_exists( $path ) )
+        touch( $path );
+
+    $res = realpath( $path );
```

Comment:
I don't quite understand why you first replace backward slashes with forward slashes, and later call `realpath()` which will reintroduce the backslashes.

Reply:
Because `realpath( "C:\phpdoc\en" )` will fail on Windows, I know that is because escaping and not directly related with realpath, but it really frustrating to have some literal "file part" *sometimes* behaving differently.

And to avoid code end looking like [this](https://github.com/php/doc-base/blob/master/configure.php#L30C1-L30C160): 
```
const RNG_SCHEMA_DIR = __DIR__ . DIRECTORY_SEPARATOR . 'docbook' . DIRECTORY_SEPARATOR . 'docbook-v5.2-os' . DIRECTORY_SEPARATOR . 'rng' . DIRECTORY_SEPARATOR;
```

Reply:
I cannot test ir now (soon starting some new year holidays), but I vaguely remember that libxml will accept foward slashes on Windows, backward slashes, but not mixed slashes, e.g., `C:\phpdoc/en`.

Reply:
Well, yeah, you can usually just use forward slashes on Windows. But I still don't understand why you first replace backward slashes with forward slashes, and later call realpath() which will reintroduce the backslashes. I mean that has nothing to do with literal file paths, nor with mixing forward and backward slashes.

Reply:
I will have to test again, without this PR, to see exactly what was broken. I think there is something git related, as this function was copied from `languages.php`, that does a lot of git invocations.

Reply:
Remembered now. `realpath()` does not reintroduces foward slashes in all cases. It's not even guaranteed to return string. This function is [called](https://github.com/php/doc-base/pull/207/files#diff-6c87a6e32d355c274d829beb3d1f6461d84cb02a709af1d8c9a604a5afe631c4R105) for expected non existent files, where `realpath()` returns `bool`, and I was under impression that this may fails on Windows, and at the time (as now) it's difficult for me to access Windows boxes to make tests. Yet. I remembered that forwards slashes always worked.

Also, these is some evidence this was the norm on doc-base scripts. `xml-check.php` [does it](https://github.com/php/doc-base/blob/master/scripts/xml-check.php#L32), and I remember that previous `file-entities.php` there is a lot oft this too.

Reply:
https://github.com/php/doc-base/blob/643f7dd80707bef61fa661ca6247d344fc46f4bc/scripts/xml-check.php#L32

That code calls `realpath()` and afterwards replaces the backslashes with forward slashes. That can make sense. Even if the file does not exists, you get `false` and the `str_replace()` will return an empty string. Even that might make sense.

I still fail to understand, why you first replace forward slashes with backward slashes, and then call `realpath()` which will always use backward slashes on Windows (unless the file does not exist, in which case the given filepath will be returned with forward slashes). Maybe you want something like this:

````php
{
    if ( $touch && ! file_exists( $path ) )
        touch( $path );

    $res = realpath( $path );
    if ($res !== false)
        $path = $res;

    return str_replace( "\\" , '/' , $path );
}

Reply:
> I still fail to understand, why you first replace forward slashes with backward slashes

To avoid mixed slashes (`C:\phpdoc/en`) when `realpath()` fails. All backwards, all fowards, never mixed.

> and then call `realpath()` which will always use backward slashes on Windows

The "foward slashes everywhere" refers to using foward slashes everywhere *in source code*. Having *all* slashes being foward on Windows when *used* (replaced), is better for compatibility.

That said, I will remove this line.